include/nbl/builtin/hlsl/workgroup/shared_scan.hlsl

```diff
@@ -0,0 +1,184 @@
+// Copyright (C) 2023 - DevSH Graphics Programming Sp. z O.O.
+// This file is part of the "Nabla Engine".
+// For conditions of distribution and use, see copyright notice in nabla.h
+#ifndef _NBL_BUILTIN_HLSL_WORKGROUP_SHARED_SCAN_INCLUDED_
+#define _NBL_BUILTIN_HLSL_WORKGROUP_SHARED_SCAN_INCLUDED_
+
+#include "nbl/builtin/hlsl/workgroup/broadcast.hlsl"
+#include "nbl/builtin/hlsl/glsl_compat/core.hlsl"
+#include "nbl/builtin/hlsl/glsl_compat/subgroup_basic.hlsl"
+
+namespace nbl 
+{
+namespace hlsl
+{
+namespace workgroup
+{
+
+template<typename T, class SubgroupOp, class SharedAccessor>
+struct WorkgroupScanHead
+{
+    bool isScan;
+    T identity;
+    T firstLevelScan;
+    uint itemCount;
+    uint lastInvocation;
+    uint lastInvocationInLevel;
+    uint scanStoreIndex;
+
+    static WorkgroupScanHead create(bool isScan, T identity, uint itemCount)
+    {
+        WorkgroupScanHead<T, SubgroupOp, SharedAccessor> wsh;
+        wsh.isScan = isScan;
+        wsh.identity = identity;
+        wsh.itemCount = itemCount;
+        wsh.lastInvocation = itemCount - 1u;
+        return wsh;
+    }
+
+    T operator()(T value)
+    {
+        const uint subgroupMask = glsl::gl_SubgroupSize() - 1u;
+        SharedAccessor sharedAccessor;
+        lastInvocationInLevel = lastInvocation;
+        
+        SubgroupOp subgroupOp;
+        firstLevelScan = subgroupOp(value);
+        T scan = firstLevelScan;
+        
+        const bool isLastSubgroupInvocation = glsl::gl_SubgroupInvocationID() == glsl::gl_SubgroupSize() - 1u;
+        
+        // Since we are scanning the RESULT of the initial scan (which paired one input per subgroup invocation) 
+        // every group of 64 invocations has been coallesced into 1 result value. This means that the results of 
+        // the first SubgroupSz^2 invocations will be processed by the first subgroup and so on.
+        // Consequently, those first SubgroupSz^2 invocations will store their results on SubgroupSz scratch slots 
+        // with halfSubgroupSz padding and the next level will follow the same + the previous as an `offset`.
+        const uint offset = (gl_LocalInvocationIndex >> glsl::gl_SubgroupSizeLog2()) & ~subgroupMask; // For subgroupSz = 64, first 4095 invocations get offset 0, 4096-8191 offset 64, then 128 etc.
+        const uint memBegin = offset;
+        uint nextLevelStoreIndex = memBegin + glsl::gl_SubgroupID();
+        scanStoreIndex = gl_LocalInvocationIndex;
+        
+        bool participate = gl_LocalInvocationIndex <= lastInvocationInLevel;
+        while(lastInvocationInLevel >= glsl::gl_SubgroupSize() * glsl::gl_SubgroupSize())
+        {
+            sharedAccessor.main.workgroupExecutionAndMemoryBarrier();
```

Comment:
because you're not touching the accessor for the first subgroup op, this barrier can move to the end of the loop like here https://github.com/Devsh-Graphics-Programming/Nabla/pull/519/files#r1327234386

Reply:
actually this barrier wouldn't be needed if you ping-ponged when not scanning (reducing)